node/node.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package node represents the Ethereum protocol stack container.
+package node
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sync"
+	"syscall"
+
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/p2p"
+)
+
+var (
+	ErrDatadirUsed       = errors.New("datadir already used")
+	ErrNodeStopped       = errors.New("node not started")
+	ErrNodeRunning       = errors.New("node already running")
+	ErrServiceUnknown    = errors.New("service not registered")
+	ErrServiceRegistered = errors.New("service already registered")
+
+	datadirInUseErrnos = map[uint]bool{11: true, 32: true, 35: true}
+)
+
+// Node represents a P2P node into which arbitrary services might be registered.
+type Node struct {
+	datadir string                        // Path to the currently used data directory
+	config  *p2p.Server                   // Configuration of the underlying P2P networking layer
+	stack   map[string]ServiceConstructor // Protocol stack registered into this node
+	emux    *event.TypeMux                // Event multiplexer used between the services of a stack
+
+	running  *p2p.Server        // Currently running P2P networking layer
+	services map[string]Service // Currently running services
+
+	stop chan struct{} // Channel to wait for termination notifications
+	lock sync.RWMutex
+}
+
+// New creates a new P2P node, ready for protocol registration.
+func New(conf *Config) (*Node, error) {
+	// Ensure the data directory exists, failing if it cannot be created
+	if conf.DataDir != "" {
+		if err := os.MkdirAll(conf.DataDir, 0700); err != nil {
+			return nil, err
+		}
+	}
+	// Assemble the networking layer and the node itself
+	nodeDbPath := ""
+	if conf.DataDir != "" {
+		nodeDbPath = filepath.Join(conf.DataDir, datadirNodeDatabase)
+	}
+	return &Node{
+		datadir: conf.DataDir,
+		config: &p2p.Server{
```

Comment:
Short story: No, because it's not a server, it's just the configuration for the server :P

Long story: A `p2p.Server` instance cannot be restarted. The only way to cleanly do a restart is to tear the old one down and instantiate a new one. However, to create a new version, I need the configurations of the old one. However, theere are no guarantees that the fields remain the same (they actually do not), so it is generally cleaner to start from a known good state than to hope no bad thing happened during runtime.

The actual running `p2p.Server` is called `running`.